sway-lsp/src/capabilities/document_symbol.rs

```diff
@@ -1,60 +1,499 @@
+/// the opening `{` character. Trims any trailing whitespace.
+///
+/// If the `{` character is not found, the entire string is returned without trailing whitespace.
+///
+/// # Examples
+///
+/// ```
+/// let impl_example = "impl<T> Setter<T> for FooBarData<T> {\n    fn set(self, new_value: T) -> Self {\n        FooBarData {\n            value: new_value,\n        }\n    }\n}";
+/// let result = extract_header(impl_example);
+/// assert_eq!(result, "impl<T> Setter<T> for FooBarData<T>");
+///
+/// let abi_example = "abi MyAbi : MySuperAbi {\n    fn bar();\n}";
+/// let result = extract_header(abi_example);
+/// assert_eq!(result, "abi MyAbi : MySuperAbi");
+/// ```
+fn extract_header(s: &str) -> &str {
+    if let Some(pos) = s.find('{') {
+        s[..pos].trim_end()
+    } else {
+        s.trim_end()
+    }
+}
+
+/// Builder for creating [`DocumentSymbol`] instances with method chaining.
+/// Initializes with empty name, NULL kind, and zero position ranges.
+pub struct DocumentSymbolBuilder {
+    name: String,
+    detail: Option<String>,
+    kind: lsp_types::SymbolKind,
+    tags: Option<Vec<lsp_types::SymbolTag>>,
+    range: lsp_types::Range,
+    selection_range: lsp_types::Range,
+    children: Option<Vec<DocumentSymbol>>,
+    deprecated: Option<bool>,
+}
+
+impl Default for DocumentSymbolBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DocumentSymbolBuilder {
+    pub fn new() -> Self {
+        Self {
+            name: String::new(),
+            kind: lsp_types::SymbolKind::NULL,
+            range: lsp_types::Range::new(
+                lsp_types::Position::new(0, 0),
+                lsp_types::Position::new(0, 0),
+            ),
+            selection_range: lsp_types::Range::new(
+                lsp_types::Position::new(0, 0),
+                lsp_types::Position::new(0, 0),
+            ),
+            detail: None,
+            tags: None,
+            children: None,
+            deprecated: None,
         }
-        SymbolKind::Function | SymbolKind::DeriveHelper | SymbolKind::Intrinsic => {
-            lsp_types::SymbolKind::FUNCTION
+    }
+
+    pub fn name(mut self, name: impl Into<String>) -> Self {
+        self.name = name.into();
+        self
+    }
+
+    pub fn kind(mut self, kind: lsp_types::SymbolKind) -> Self {
+        self.kind = kind;
+        self
+    }
+
+    pub fn range(mut self, range: lsp_types::Range) -> Self {
+        self.range = range;
+        self
+    }
+
+    pub fn selection_range(mut self, range: lsp_types::Range) -> Self {
+        self.selection_range = range;
+        self
+    }
+
+    pub fn detail(mut self, detail: Option<String>) -> Self {
+        self.detail = detail;
+        self
+    }
+
+    pub fn tags(mut self, tags: Vec<lsp_types::SymbolTag>) -> Self {
+        self.tags = Some(tags);
+        self
+    }
+
+    pub fn children(mut self, children: Vec<DocumentSymbol>) -> Self {
+        self.children = Some(children);
+        self
+    }
+
```

Comment:
These `Option<Vec>` types start off default `None` but then are set to `Some(Vec)` without any way of being set back to `None`. Should these setters take in an `Option<Vec>` instead?

Reply:
The current design is intentional for the builder pattern. While it's true that we can't set these fields back to `None` after setting them to `Some(Vec)`, this is actually a common and beneficial approach for builders because builders are typically used in a single chain of method calls to construct an object, not to repeatedly modify an existing instance. The goal is to provide a clear, one-way path to construct the final object.